### PR TITLE
NE-356: Add HAProxy config files to must-gather

### PR DESCRIPTION
**Pull request by sgreene570**
Add script file collection-scripts/gather_haproxy_config to collect haproxy config files for each ingress controller in the cluster.

This is in support of https://issues.redhat.com/browse/NE-356.

### CODE REVIEW

File: collection-scripts/gather_haproxy_config
Comment (Miciah) [inline collection-scripts/gather_haproxy_config:13]: Just a suggestion:  You could use `-o name`, although that has the disadvantage that you need to strip off the leading resource qualifier (e.g., `oc get ingresscontrollers -o name` prints "ingresscontroller.operator.openshift.io/default", and you want just "default"):
```suggestion
    INGRESS_CONTROLLERS=$(oc get ingresscontroller -n openshift-ingress-operator -o name)
    for IC in ${INGRESS_CONTROLLERS}; do
        PODS=($(oc get pods -n openshift-ingress -o name -l ingresscontroller.operator.openshift.io/deployment-ingresscontroller="${IC#*/}"))
```

File: collection-scripts/gather_haproxy_config
Comment (Miciah) [inline collection-scripts/gather_haproxy_config:15]: It might be useful to have `haproxy.config` from each pod in some cases.

Comment (Miciah) [inline collection-scripts/gather_haproxy_config:15]: You did `mkdir -p "${HAPROXY_CONFIG_PATH}"/` earlier, but you didn't `mkdir "${HAPROXY_CONFIG_PATH}"/"${IC}"/`...  But all right, it looks like `oc` takes care of making the directory: https://github.com/openshift/oc/blob/1bd1f5ff4aa98d6e4184c2948928b6111fa99191/vendor/k8s.io/kubectl/pkg/cmd/cp/cp.go#L466
So I think this is fine.

File: collection-scripts/gather_haproxy_config
Comment (sgreene570) [reply collection-scripts/gather_haproxy_config:13]: Other scripts in `must-gather` seem to more often use the `columns=...` approach, so i'll stick with the current approach for consistency sake. That being said, I will wrap the line in `()` to assign the values as an array. Thanks!

File: collection-scripts/gather_haproxy_config
Comment (sgreene570) [reply collection-scripts/gather_haproxy_config:15]: Right, I was going back on fourth on this. But you are correct, in the event of 1/n routers becoming wedged, the configs may be different (and having all of them would help with debugging this case).

Comment (sgreene570) [reply collection-scripts/gather_haproxy_config:15]: Yea, I noticed the oc command did the directory creations for me, so I decided that I didn't need to explicitly make them.

Comment (sgreene570) [reply collection-scripts/gather_haproxy_config]: `7d138e0` gathers each pod's HAProxy config using a for loop.

**Comment by davemulford** [comment]
@sferich888 is this something we want to happen for every must-gather that gets run?

**Comment by sgreene570** [comment]
> @sferich888 is this something we want to happen for every must-gather that gets run?

Any particular reason we would not want this to run for every must-gather? Do you have particular concerns in mind?

**Review by knobunc — APPROVED** [review]
/approve

File: collection-scripts/gather_haproxy_config
Comment (sferich888) [inline collection-scripts/gather_haproxy_config:14]: In the 3.x days (and now even) we also collected, things that affected the control loop. Should we also collect those here? 

https://access.redhat.com/solutions/1542293#routing

File: collection-scripts/gather_haproxy_config
Comment (sferich888) [inline collection-scripts/gather_haproxy_config:12]: Should we thread this (try an make all these pod collection calls in parallel vs serial)?

File: collection-scripts/gather_haproxy_config
Comment (sferich888) [inline collection-scripts/gather_haproxy_config:4]: Do we want to put this under, ingress-controllers? What if we move away from haproxy in the future?

File: collection-scripts/gather_haproxy_config
Comment (sferich888) [inline collection-scripts/gather_haproxy_config:13]: Do we need to make this loop configurable, what happens if the ingress-controller pod is not an haproxy pod? Do we need to make this configurable or integrated with the pod (some data collection convention), so that this works with 3rd party ingress-controllers?

File: collection-scripts/gather
Comment (sferich888) [inline collection-scripts/gather:61]: Do we want this collected by default?

**Review by sferich888 — APPROVED** [review]
I think this is fine (in general), but I wonder if we should consider a few possible improvements/alterations (possibly for a later date/time).

File: collection-scripts/gather_haproxy_config
Comment (sgreene570) [reply collection-scripts/gather_haproxy_config:13]: 3rd party ingress controllers wouldn't be able to make use of the OCP ingress controller API, and this scripts selects `PODS` by OCP ingress controller, so I think we are OK here.

File: collection-scripts/gather_haproxy_config
Comment (sgreene570) [reply collection-scripts/gather_haproxy_config:14]: I believe must-gather already collects everything else for the router pods that would make sense to have.

**Comment by sgreene570** [comment]
Pushing some fixes shortly
/hold 
(since this already has lgtm and approved tags)

File: collection-scripts/gather_haproxy_config
Comment (sgreene570) [reply collection-scripts/gather_haproxy_config:12]: Fixed.

File: collection-scripts/gather_haproxy_config
Comment (sgreene570) [reply collection-scripts/gather_haproxy_config]: change to `ingress_controllers`

File: collection-scripts/gather
Comment (sgreene570) [reply collection-scripts/gather:61]: Any reason not to? Do you feel that this should not be collected by default?

Comment (sferich888) [reply collection-scripts/gather:61]: I don't like collecting things in this manner (by default), but .. in this situations I don't think we have a better solution.

File: collection-scripts/gather_haproxy_config
Comment (sferich888) [reply collection-scripts/gather_haproxy_config:12]: Thanks.

**Pull request merged**
